core/src/main/java/org/mule/runtime/core/internal/processor/strategy/AbstractCachedThreadReactorSinkProvider.java

```diff
@@ -51,9 +72,43 @@ protected void invalidateAll() {
   public FluxSink<CoreEvent> getSink() {
     TransactionCoordination txCoord = TransactionCoordination.getInstance();
     if (txCoord.runningNestedTransaction()) {
-      return sinksNestedTx.get(txCoord.getTransaction(), tx -> createSink());
+      if (sinkIndexEnabled) {
+        return getNestedFluxSinkWrapper(txCoord);
+      } else {
+        return legacySinksNestedTx.get(txCoord.getTransaction(), tx -> new FluxSinkWrapper(createSink()));
+      }
     } else {
-      return sinks.get(currentThread(), t -> createSink());
+      if (sinkIndexEnabled) {
+        return getSimpleFluxSinkWrapper();
+      } else {
+        return legacySinks.get(currentThread(), t -> new FluxSinkWrapper(createSink()));
+      }
+    }
+  }
+
+  private FluxSink<CoreEvent> getNestedFluxSinkWrapper(TransactionCoordination txCoord) {
+    int index = 0;
+    while (true) {
+      FluxSinkWrapper fluxSinkWrapper =
+          sinksNestedTx.get(txCoord.getTransaction().getId() + "-" + index, t -> new FluxSinkWrapper(createSink()));
```

Comment:
can the transacion id be repeated? form what i see in the code it is a uuid

Reply:
I don't think it would be an issue if it eventually repeats, as long as we can find a sink that is not being used we can go ahead and use it I guess.
If I'm not getting it wrong, the transaction ID (as well as the thread ID in the other case) are only used for bucketing to make it less likely to find sinks that are in use, hence reducing the number of cache lookups.
I'm even going to throw into the board the possibility to just use the Thread ID here as well... just for simplifying things. It may also increase the chance of being able to reuse an already existing sink and avoid creating a new one.
It looks like it comes down to a trade-off between the likelihood of being able to reuse an existing Sink and the time it may take to find an available one.
That being said, it is a good idea to make sure we don't steal a Sink that might be chosen from a `getSink` being executed in a different thread, because the `isBeingUsed` check is not synchronized with effectively marking the Sink as used before handing it over.

Reply:
In this case, we always use the sink from the same thread (even for nested transactions) because in the case of transactions no matter if they are nested or not everything executes in the same thread. That's why the isBeingUsed is not synchronized.

Reply:
Yes, I know, I was just pointing it out in case we decided to do something different in term of bucketing

Reply:
The transactionId is used to guarantee that no work stealing is done inside reactor. If there is work stealing, the transactions that are handled through thread locals can be mixed up.

If the same sink is used and the next is executed from two different threads, after one thread executes everything, it can go on executing the chain for the event that the other thread enqueued (and it will not have the correct transaction information in the thread local). I think that was the original problem tackled by this class.